lib/src/builder/parsing/version.dart

```diff
@@ -0,0 +1,207 @@
+  static const neutral = 10;
+
+  /// Likely to be considered a given version.
+  static const likely = 100;
+
+  /// Certain to be considered a given version.
+  static const certain = 100000;
+}
+
+/// The current versions of the boilerplate that the builder detects.
+///
+/// Related: [resolveVersion]
+enum Version {
+  // "v1" is for the transformer-based syntax :)
+  v2_legacyBackwardsCompat,
+  v3_legacyDart2Only,
+  v4_mixinBased,
+}
+
+/// Priority of the possible boilerplate versions.
+///
+/// In the event that two different [Version]s have the same confidence score,
+/// version of high priority wins out.
+///
+/// The length of this iterable should match the length of values in [Version].
+const _versionsInPriorityOrder = [
+  Version.v4_mixinBased,
+  Version.v2_legacyBackwardsCompat,
+  Version.v3_legacyDart2Only,
+];
+
+/// Utils to add capabilities to the [Version] enum.
+extension VersionExtension on Version {
+  /// Validates whether or not the current version is one grouped in the "Legacy" category.
+  ///
+  /// Because each boilerplate member and declaration has a version property,
+  /// this getter provides a declarative way to detect if that object is legacy.
+  bool get isLegacy =>
+      this == Version.v2_legacyBackwardsCompat || this == Version.v3_legacyDart2Only;
+}
+
+/// An encapsulation of the data useful when evaluating a member version.
+///
+/// This can also be used when the builder is making generation decisions.
+class VersionConfidencePair {
+  /// The evaluated version for a group of members.
+  ///
+  /// __Note:__ this is not always the version of an individual member, but rather
+  /// can be the version evaluation that is used to generate the class implementations.
+  /// This is the case then the object is being returned from [resolveVersion].
+  final Version version;
+
+  /// The numerical representation of version confidence.
+  ///
+  /// When being returned from [resolveVersion], this is the total confidence
+  /// for all included members summed and then averaged.
+  final num confidence;
+
+  VersionConfidencePair(this.version, this.confidence);
+
+  /// Whether or not there is enough confidence to generate an implementation.
+  ///
+  /// Based on the boilerplate members cumulatively, there may not be enough data
+  /// to signify that generation is necessary. This is the case when the [confidence]
+  /// (the average confidence score for all members) is not above [Confidence.neutral].
+  ///
+  /// __Note:__ Since the purpose of this getter is specific to the generator, it should
+  /// not be trusted unless this is initialized as the return value for [resolveVersion].
+  bool get shouldGenerate => confidence > Confidence.neutral;
+}
+
+/// Returns a confidence pair with
+/// - a [VersionConfidencePair.version] equal to the highest-confidence version among all [members]
+/// - a [VersionConfidencePair.confidence] equal to the average of all confidences for that version
```

Comment:
What happens if the guessed version is wrong? Or is that not possible?

Reply:
That shouldn't be possible provided that the confidences are accurate; the guessed version should represent the most likely version given the information available.

That being said, it's possible for this to differ from the consumer's intended version if they mess it up enough, but there's not much we can do about that.